include/graaflib/algorithm/topological_sorting.tpp

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <graaflib/algorithm/cycle_detection.h>
+#include <graaflib/algorithm/topological_sorting.h>
+
+#include <vector>
+#include <unordered_map>
+
+namespace graaf::algorithm {
+
+namespace detail {
+
+enum class vertex_status { UNVISITED, VISITED};
+
+// DFS topological sort
+template <typename V, typename E>
+void do_dfs_topological_sort(
+    const graph<V, E, graph_type::DIRECTED>& graph,
+    vertex_id_t start_vertex,
+    std::unordered_map<vertex_id_t, vertex_status>& processed_vertices,
+    std::vector<vertex_id_t>& sorted_vertices) {
+    
+    processed_vertices[start_vertex] = vertex_status::VISITED;
+    for (const auto& next_vertex : graph.get_neighbors(start_vertex)) {
+        if (processed_vertices[next_vertex] == vertex_status::UNVISITED) {
+            do_dfs_topological_sort(graph, next_vertex, processed_vertices,
+                sorted_vertices);
+        }
+    }
+
+        sorted_vertices.push_back(start_vertex);
```

Comment:
If you add the `start_vertex` to the `sorted_vertices` before the `for`-loop, do you already have the vertices recorded in the correct order? Maybe it would be possible to get rid of the `std::reverse` then.

Reply:
I'm not sure about that. In this case, vertices will be stored as usual in DFS traversal, ignoring all other dependencies.
Consider this directed graph
       0
    /   \
   *     *
   1    2
   \    /
    *  *
      3
If we do preorder, the vertices will be 0–1–3–2, which is incorrect for TS.
But if we do post order 0 - 1 - 2 - 3, Or maybe I missed something, correct me if I'm getting something wrong.

Reply:
I quickly checked @bobluppes's proposal and it indeed fails the tests. I would suggest going with the solution in this PR and think about ways getting rid of the `std::reverse` in a follow up.

Reply:
Ah, my bad indeed. One thing we could investigate is to use a `std::deque` rather than a `vector` and call `push_front`. However, let's not optimize prematurely and let's tackle this in a follow up.